builder/dockerfile/builder.go

```diff
@@ -168,62 +173,69 @@ func emitImageID(aux *streamformatter.AuxFormatter, state *dispatchState) error
 	}
 	return aux.Emit(types.BuildResult{ID: state.imageID})
 }
+func convertMapToEnvs(m map[string]string) []string {
+	result := []string{}
+	for k, v := range m {
+		result = append(result, k+"="+v)
+	}
+	return result
+}
+func (b *Builder) dispatchDockerfileWithCancellation(parseResult []instructions.BuildableStage, metaArgs []instructions.ArgCommand, escapeToken rune, source builder.Source) (*dispatchState, error) {
+	var stageBuilder *stageBuilder
 
-func (b *Builder) dispatchDockerfileWithCancellation(dockerfile *parser.Result, source builder.Source) (*dispatchState, error) {
-	shlex := NewShellLex(dockerfile.EscapeToken)
-	state := newDispatchState()
-	total := len(dockerfile.AST.Children)
-	var err error
-	for i, n := range dockerfile.AST.Children {
-		select {
-		case <-b.clientCtx.Done():
-			logrus.Debug("Builder: build cancelled!")
-			fmt.Fprint(b.Stdout, "Build cancelled")
-			buildsFailed.WithValues(metricsBuildCanceled).Inc()
-			return nil, errors.New("Build cancelled")
-		default:
-			// Not cancelled yet, keep going...
+	totalCommands := len(metaArgs)
+	currentCommandIndex := 1
+	for _, stage := range parseResult {
+		totalCommands += len(stage.Commands)
+	}
+	for _, meta := range metaArgs {
+		shlex := NewShellLex(escapeToken)
+		envs := convertMapToEnvs(b.buildArgs.GetAllAllowed())
+		if err := meta.Expand(func(word string) (string, error) {
+			return shlex.ProcessWord(word, envs)
+		}); err != nil {
+			return nil, err
 		}
+		b.buildArgs.AddArg(meta.Name, meta.Value)
+		b.buildArgs.AddMetaArg(meta.Name, meta.Value)
 
-		// If this is a FROM and we have a previous image then
-		// emit an aux message for that image since it is the
-		// end of the previous stage
-		if n.Value == command.From {
-			if err := emitImageID(b.Aux, state); err != nil {
-				return nil, err
+		fmt.Fprintf(b.Stdout, "%v / %v %v", currentCommandIndex, totalCommands, &meta)
+		currentCommandIndex++
+		fmt.Fprintln(b.Stdout)
+	}
+
+	for _, stage := range parseResult {
+		stageBuilder = newStageBuilder(b, escapeToken, source)
+		for _, cmd := range stage.Commands {
+			select {
+			case <-b.clientCtx.Done():
+				logrus.Debug("Builder: build cancelled!")
+				fmt.Fprint(b.Stdout, "Build cancelled")
+				buildsFailed.WithValues(metricsBuildCanceled).Inc()
+				return nil, errors.New("Build cancelled")
+			default:
+				// Not cancelled yet, keep going...
 			}
-		}
 
-		if n.Value == command.From && state.isCurrentStage(b.options.Target) {
-			break
-		}
+			fmt.Fprintf(b.Stdout, "%v / %v %v", currentCommandIndex, totalCommands, cmd)
```

Comment:
"Step %d/%d : %v" for compatibility?

I think there are user applications that depend on the current logging format.

(should be converted to machine-readable format in the new RPC)